File: src/tok/charsmap.zig
//! charsmap.zig — SentencePiece's `nmt_nfkc`, read from the artifact.
//!
//! The `.spm` vocabulary carries a `normalizer_spec` naming `nmt_nfkc` and a
//! `precompiled_charsmap`: a darts-clone double-array trie over UTF-8 byte
//! sequences followed by a pool of NUL-terminated replacements. Matching is
//! longest-prefix; a hit emits its replacement, a miss copies one UTF-8
//! character through.
//!
//! **This is not an implementation of NFKC.** It is an interpreter for the
//! table the model was trained with. Unicode normalization has a specification,
//! several revisions of it, and the model does not care about any of them — it
//! cares about the 237 KB of rewrite rules baked into its own vocabulary. ADR
//! 0017.
//!
//! Layout, little-endian throughout:
//!
//!     u32   trie_bytes
//!     [..]  trie_bytes of darts-clone units, u32 each
//!     [..]  replacement pool, NUL-terminated

const std = @import("std");
const assert = std.debug.assert;

pub const Error = error{OutTooSmall};

/// A matched rewrite rule: where its replacement lives in the pool, and how
/// many input bytes it consumes.
const Match = struct { at: u32, len: u32 };

/// A charsmap the artifact did not ship. `normalize` copies through.
pub const none: Charsmap = .{ .blob = &.{} };

pub const Charsmap = struct {
    blob: []const u8,

    pub fn present(self: Charsmap) bool {
        return self.blob.len >= 4;
    }

    fn trieLen(self: Charsmap) u32 {
        assert(self.present());
        return std.mem.readInt(u32, self.blob[0..4], .little) / 4;
    }

    fn unit(self: Charsmap, i: u32) u32 {
        const at = 4 + @as(usize, i) * 4;
        assert(at + 4 <= self.blob.len);
        return std.mem.readInt(u32, self.blob[at..][0..4], .little);
    }

    fn pool(self: Charsmap) []const u8 {
        const at = 4 + @as(usize, self.trieLen()) * 4;
        assert(at <= self.blob.len);
        return self.blob[at..];
    }

    /// Longest rewrite rule matching the head of `key`. Returns the pool offset
    /// and how many input bytes it consumes, or null when no rule applies.
    ///
    /// The bit layout is darts-clone's: bit 8 marks a leaf, bits 10.. hold the
    /// child offset scaled by bit 9, and the low byte is the edge label.
    fn longest(self: Charsmap, key: []const u8) ?Match {
        assert(self.present());
        const n = self.trieLen();
        if (n == 0) return null;

        var node: u32 = 0;
        var u = self.unit(node);
        node ^= offsetOf(u);

        var best: ?Match = null;
        for (key, 0..) |b, i| {
            node ^= b;
            if (node >= n) return best;
            u = self.unit(node);
            if (u & 0x800000FF != b) return best;
            node ^= offsetOf(u);
            if (node >= n) return best;
            if ((u >> 8) & 1 == 1) {
                best = .{ .at = self.unit(node) & 0x7FFFFFFF, .len = @intCast(i + 1) };
            }
        }
        return best;
    }

    fn offsetOf(u: u32) u32 {
        return (u >> 10) << @intCast((u & 0x200) >> 6);
    }

    fn replacement(self: Charsmap, at: u32) []const u8 {
        const p = self.pool();
        assert(at < p.len);
        const rest = p[at..];
        const end = std.mem.indexOfScalar(u8, rest, 0) orelse rest.len;
        return rest[0..end];
    }

    /// Rewrites `src` into `out`, returning the bytes written.
    ///
    /// `error.OutTooSmall` rather than truncation: the worst rule in the
    /// shipped tables expands 3 bytes to 33 (U+FDFA), so an adversarial input
    /// can grow 11x, and a half-normalized string is a wrong string rather than
    /// a short one. The caller turns this into `src_too_long`, which is what it
    /// is — after normalization, the text no longer fits.
    pub fn normalize(self: Charsmap, src: []const u8, out: []u8) Error!u32 {
        assert(out.len <= std.math.maxInt(u32));

        if (!self.present()) {
            if (src.len > out.len) return error.OutTooSmall;
            @memcpy(out[0..src.len], src);
            return @intCast(src.len);
        }

        var r: usize = 0;
        var w: u32 = 0;
        // SPEC §12.3: bounded. Every arm advances `r` by at least one byte.
        for (0..src.len) |_| {
            if (r >= src.len) break;
            if (self.longest(src[r..])) |hit| {
                const rep = self.replacement(hit.at);
                if (w + rep.len > out.len) return error.OutTooSmall;
                @memcpy(out[w..][0..rep.len], rep);
                w += @intCast(rep.len);
                r += hit.len;
                assert(hit.len > 0);
                continue;
            }
            const step = utf8Len(src[r]);
            const take = @min(step, src.len - r);
            if (w + take > out.len) return error.OutTooSmall;
            @memcpy(out[w..][0..take], src[r..][0..take]);
            w += @intCast(take);
            r += take;
        }

        assert(r == src.len);
        return w;
    }
};

fn utf8Len(b: u8) usize {
    if (b >= 0xF0) return 4;
    if (b >= 0xE0) return 3;
    if (b >= 0xC0) return 2;
    return 1;
}

test "an absent charsmap copies through" {
    var buf: [32]u8 = undefined;
    const n = try none.normalize("hola", &buf);
    try std.testing.expectEqualStrings("hola", buf[0..n]);
}

test "an absent charsmap still refuses to overflow" {
    var buf: [2]u8 = undefined;
    try std.testing.expectError(error.OutTooSmall, none.normalize("hola", &buf));
}

test "utf8Len covers every lead byte class" {
    try std.testing.expectEqual(@as(usize, 1), utf8Len('a'));
    try std.testing.expectEqual(@as(usize, 2), utf8Len(0xC3));
    try std.testing.expectEqual(@as(usize, 3), utf8Len(0xE2));
    try std.testing.expectEqual(@as(usize, 4), utf8Len(0xF0));
}

File: tools/charsmap.py
#!/usr/bin/env python3
"""charsmap.py — SentencePiece's `nmt_nfkc` normalizer, read from the artifact.

The `.spm` vocabulary carries a `normalizer_spec` naming `nmt_nfkc` and a
`precompiled_charsmap`: a darts-clone double-array trie over UTF-8 byte
sequences, followed by a pool of NUL-terminated replacement strings. Matching
is longest-prefix; a hit appends its replacement, a miss copies one UTF-8
character through.

This is the reference implementation, used by the converter to build the
runtime's table and by the tests to check `src/tok/charsmap.zig` against it.
Nothing here is a reimplementation of NFKC: the mapping is whatever the model
was trained with, read out of the model.

Layout:

    u32le trie_size
    trie_size bytes   darts-clone units, u32le each
    remainder         replacement pool, NUL-terminated strings
"""

from __future__ import annotations

import struct
from pathlib import Path


def _varint(buf: bytes, i: int) -> tuple[int, int]:
    v = shift = 0
    while True:
        b = buf[i]
        i += 1
        v |= (b & 0x7F) << shift
        shift += 7
        if not b & 0x80:
            return v, i


def read_spec(spm_path: Path) -> tuple[str, bytes]:
    """Returns (normalizer name, precompiled_charsmap) from a .spm file."""
    raw = spm_path.read_bytes()
    i = 0
    while i < len(raw):
        tag, i = _varint(raw, i)
        ln, i = _varint(raw, i)
        if tag >> 3 == 3:  # normalizer_spec
            sub, j, name, charsmap = raw[i:i + ln], 0, "", b""
            while j < len(sub):
                t, j = _varint(sub, j)
                sl, j = _varint(sub, j)
                if t >> 3 == 1:
                    name = sub[j:j + sl].decode()
                elif t >> 3 == 2:
                    charsmap = sub[j:j + sl]
                j += sl
            return name, charsmap
        i += ln
    return "", b""


class Charsmap:
    """Longest-prefix rewriter over a darts-clone trie."""

    def __init__(self, blob: bytes):
        (trie_size,) = struct.unpack_from("<I", blob, 0)
        assert trie_size % 4 == 0, "trie blob is not a whole number of units"
        self.units = list(struct.unpack_from(f"<{trie_size // 4}I", blob, 4))
        self.pool = blob[4 + trie_size:]

    # darts-clone unit accessors, from `darts.h`.
    @staticmethod
    def _has_leaf(u: int) -> bool:
        return bool((u >> 8) & 1)

    @staticmethod
    def _value(u: int) -> int:
        return u & 0x7FFFFFFF

    @staticmethod
    def _label(u: int) -> int:
        return u & 0x800000FF

    @staticmethod
    def _offset(u: int) -> int:
        return (u >> 10) << ((u & 0x200) >> 6)

    def longest(self, key: bytes) -> tuple[int, int]:
        """Longest match at the head of `key`. Returns (pool offset, bytes
        consumed), or (-1, 0) when nothing matches."""
        node = 0
        unit = self.units[node]
        node ^= self._offset(unit)
        best_value, best_len = -1, 0
        for i, b in enumerate(key):
            node ^= b
            unit = self.units[node]
            if self._label(unit) != b:
                break
            node ^= self._offset(unit)
            if self._has_leaf(unit):
                best_value, best_len = self._value(self.units[node]), i + 1
        return best_value, best_len

    def replacement(self, offset: int) -> bytes:
        end = self.pool.index(b"\0", offset)
        return self.pool[offset:end]

    def normalize(self, text: bytes) -> bytes:
        out, i, n = bytearray(), 0, len(text)
        while i < n:
            value, length = self.longest(text[i:])
            if length > 0:
                out += self.replacement(value)
                i += length
                continue
            # No rule: copy one whole UTF-8 character through unchanged.
            step = 1
            b = text[i]
            if b >= 0xF0:
                step = 4
            elif b >= 0xE0:
                step = 3
            elif b >= 0xC0:
                step = 2
            out += text[i:i + step]
            i += step
        return bytes(out)


if __name__ == "__main__":
    import sys

    name, blob = read_spec(Path(sys.argv[1]))
    cm = Charsmap(blob)
    print(f"  normalizer {name}, charsmap {len(blob)} bytes, "
          f"{len(cm.units)} trie units, {len(cm.pool)} bytes of replacements")
    for line in sys.stdin:
        raw = line.rstrip("\n").encode("utf-8")
        got = cm.normalize(raw)
        flag = "" if got == raw else "  CHANGED"
        print(f"    {got.decode('utf-8', 'replace')}{flag}")
